src/services/ContributingDeveloperAuditService/scm/GitHub/api/GitHubApiClient.ts

```diff
@@ -0,0 +1,218 @@
+import { sleep } from "../../../../../utilities";
+
+interface IHttpRequestParameters {
+  baseUri: string;
+  githubPAT: string;
+}
+
+interface IHttpClientParameters extends IHttpRequestParameters {
+  apiClientName: string;
+}
+
+export interface GitHubOrganization {
+  login: string;
+}
+
+export interface GitHubRepository {
+  id: number;
+  name: string;
+  full_name: string;
+  private: boolean;
+  owner: GitHubOrganization;
+}
+
+export interface ContributingDeveloper {
+  username: string;
+  repositories: ContributingDeveloperRepositories[];
+}
+
+interface ContributingDeveloperRepositories {
+  id: number;
+  name: string;
+  lastCommit: string;
+  isPrivate: boolean;
+}
+
+export interface Commits {
+  commit: {
+    author: Author;
+  };
+}
+
+export interface Author {
+  name: string;
+  email: string;
+  date: string;
+}
+
+export const threeMonthsDate = (() => {
+  const d = new Date();
+  d.setDate(d.getDate() - 90);
+  return `${d.getUTCFullYear()}-${
+    d.getMonth() + 1
+  }-${d.getUTCDate()}T${d.getHours()}:${d.getMinutes()}:${d.getSeconds()}Z`;
+})();
+
+class GitHubApiClient {
+  private readonly client: AxiosInstance;
+
+  constructor(githubPAT: string, baseUri: string = "https://api.github.com") {
+    this.client = GitHubApiClient.createHttpClient({
+      baseUri,
+      githubPAT,
+      apiClientName: "GitHub API",
+    });
+  }
+
+  private static createHttpClient({ baseUri, githubPAT, apiClientName }: IHttpClientParameters) {
+    const client = axios.create({
+      baseURL: baseUri,
+      headers: {
+        accept: "application/vnd.github+json",
+        Authorization: `Bearer ${githubPAT}`,
+      },
+    });
+
+    client.interceptors.request.use(
+      (request) => {
+        if (request.data) {
+          soosLogger.verboseDebug(
+            apiClientName,
+            `Request URL: ${request.method?.toLocaleUpperCase()} ${request.url}`,
+          );
+          if (request.params) {
+            soosLogger.verboseDebug(
+              apiClientName,
+              `Request Params: ${JSON.stringify(request.params)}`,
+            );
+          }
+          soosLogger.verboseDebug(apiClientName, `Request Body: ${JSON.stringify(request.data)}`);
+        }
+        return request;
+      },
+      (rejectedRequest) => {
+        return Promise.reject(rejectedRequest);
+      },
+    );
+
+    client.interceptors.response.use(
+      async (response) => {
+        soosLogger.verboseDebug(apiClientName, `Response Body: ${JSON.stringify(response.data)}`);
+        if (response.config.url?.includes("per_page")) {
+          return await GitHubApiClient.handlePagination(response, client);
+        }
+        return response;
+      },
+      async (error) => {
+        const { config, response } = error;
+        const maxRetries = 3;
+        config.retryCount = config.retryCount || 0;
+
+        if (response?.status === 429 && config.retryCount < maxRetries) {
```

Comment:
If rate limited by the github api wait 3 minutes and try again, max 3 tries, but probably this could be only retried once. Also might want to change the message to include something like `If the issue persists contact support trough support@soos.io`

Reply:
they might tell you when to retry in a header of the response

Reply:
refactored this to use their headers